src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityLivingBase.java

```diff
@@ -845,6 +861,59 @@ private void causeTrackDeathUpdate(EntityLivingBase entityLivingBase) {
         }
     }
 
+    private EquipmentSlotFabric inventory = new EquipmentSlotFabric(this);
+    private EnumMap<EntityEquipmentSlot, SlotLens<Living, ItemStack>> slotLens = new EnumMap<>(EntityEquipmentSlot.class);
+
+    @Inject(method = "onUpdate", locals = LocalCapture.CAPTURE_FAILHARD,
+            at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/EntityLivingBase;getItemStackFromSlot(Lnet/minecraft/inventory/EntityEquipmentSlot;)Lnet/minecraft/item/ItemStack;"))
+    private void onGetItemStackFromSlot(CallbackInfo ci, int i, EntityEquipmentSlot[] slots, int j, int k,
+                                        EntityEquipmentSlot entityEquipmentSlot, ItemStack before) {
+        ItemStack after = this.getItemStackFromSlot(entityEquipmentSlot);
+        EntityLivingBase entity = EntityUtil.toNative(this);
+        if (!ItemStack.areItemStacksEqual(after, before)) {
+            InventoryAdapter<?, ItemStack> slotAdapter;
+            if (entity instanceof EntityPlayerMP) {
+                SlotLens<IInventory, ItemStack> slotLens;
+                IMixinInventoryPlayer inventory = (IMixinInventoryPlayer) ((EntityPlayerMP) entity).inventory;
+                PlayerInventoryLens inventoryLens = (PlayerInventoryLens) inventory.getRootLens();
+                switch (entityEquipmentSlot) {
+                case OFFHAND:
+                    slotLens = inventoryLens.getOffhandLens();
+                    break;
+                case MAINHAND:
+                    HotbarLens<IInventory, ItemStack> hotbarLens = inventoryLens.getMainLens().getHotbar();
+                    slotLens = hotbarLens.getSlot(hotbarLens.getSelectedSlotIndex(inventory.getFabric()));
+                    break;
+                default:
+                    slotLens = inventoryLens.getEquipmentLens().getSlot(entityEquipmentSlot.getIndex());
+                }
+                slotAdapter = slotLens.getAdapter(inventory.getFabric(), inventory);
+            } else {
+                if (this.slotLens.isEmpty()) {
+                    for (EntityEquipmentSlot slot : EntityEquipmentSlot.values()) {
+                        this.slotLens.put(slot, new SlotLensImpl<>(slot.getSlotIndex()));
+                    }
+                }
+                slotAdapter = this.slotLens.get(entityEquipmentSlot).getAdapter(this.inventory, null);
+            }
+            ChangeEntityEquipmentEvent event = SpongeCommonEventFactory.callChangeEntityEquipmentEvent(entity,
```

Comment:
For non-players this is missing the cause.

Reply:
Solved in 210d5dc36ab35f20e858e39f1f9a64d6d5ec7886